batch/test/test_batch.py

```diff
@@ -1777,3 +1793,417 @@ def test_region(client: BatchClient):
+    for _ in range(MAX_JOB_GROUP_NESTING_DEPTH - 1):
+        jg = jg.create_job_group()
+    jg.create_job(DOCKER_ROOT_IMAGE, ['true'])
+    b.submit()
+    status = b.wait()
+    assert status['state'] == 'success', str(b.debug_info())
+
+
+def test_cancellation_does_not_propogate_up(client: BatchClient):
+    b = create_batch(client)
+    jg = b.create_job_group()
+    job_groups = [jg]
+    for _ in range(MAX_JOB_GROUP_NESTING_DEPTH - 1):
+        jg = jg.create_job_group()
+        job_groups.append(jg)
+    b.submit()
+    job_groups[-1].cancel()
+
+    for jg in job_groups[:-1]:
+        status = jg.status()
+        assert status['state'] != 'cancelled', str(jg.debug_info())
+
+
+def test_maximum_nesting_level(client: BatchClient):
+    b = create_batch(client)
+    jg = b.create_job_group()
+    for _ in range(MAX_JOB_GROUP_NESTING_DEPTH + 1):
+        jg = jg.create_job_group()
+    with pytest.raises(httpx.ClientResponseError, match='job group exceeded the maximum level of nesting'):
+        b.submit()
+
+
+def test_all_nested_job_groups_end_up_with_correct_number_of_job_states(client: BatchClient):
+    b = create_batch(client)
+
+    jg = b.create_job_group()
+    jg.create_job(DOCKER_ROOT_IMAGE, ['true'])
+    jg.create_job(DOCKER_ROOT_IMAGE, ['false'])
+
+    job_groups = [jg]
+    for _ in range(MAX_JOB_GROUP_NESTING_DEPTH - 1):
+        jg = jg.create_job_group()
+        job_groups.append(jg)
+        jg.create_job(DOCKER_ROOT_IMAGE, ['true'])
+        jg.create_job(DOCKER_ROOT_IMAGE, ['false'])
+
+    b.submit()
+    b.wait()
+
+    n_job_groups = len(job_groups)
+    for level, jg in enumerate(job_groups):
+        status = jg.status()
+        assert status['n_succeeded'] == n_job_groups - level, str(jg.debug_info())
+        assert status['n_failed'] == n_job_groups - level, str(jg.debug_info())
+
+
+def test_cancel_job_group_with_different_updates(client: BatchClient):
+    b = create_batch(client)
+    jg = b.create_job_group()
+    j1 = jg.create_job(DOCKER_ROOT_IMAGE, ['sleep', '300'])
+    b.submit()
+
+    j2 = jg.create_job(DOCKER_ROOT_IMAGE, ['sleep', '300'])
+    b.submit()
+
+    j1._wait_for_states('Running')
+    j2._wait_for_states('Running')
+
+    jg.cancel()
+    b_status = b.wait()
+    jg_status = jg.status()
+
+    assert b_status['state'] == 'cancelled', str(b_status)
+    assert jg_status['state'] == 'cancelled', str(jg_status)
+
+    assert j1.status()['state'] == 'Cancelled', str(j1.status())
+    assert j2.status()['state'] == 'Cancelled', str(j2.status())
+
+
+def test_cancel_job_group_with_different_inst_colls(client: BatchClient):
+    b = create_batch(client)
+    jg = b.create_job_group()
+    j1 = jg.create_job(DOCKER_ROOT_IMAGE, ['sleep', '300'], resources={'memory': 'lowmem'})
+    b.submit()
+
+    j2 = jg.create_job(DOCKER_ROOT_IMAGE, ['sleep', '300'], resources={'memory': 'standard'})
+    b.submit()
+
+    j1._wait_for_states('Running')
+    j2._wait_for_states('Running')
```

Comment:
I'm worried this test is too racy because a lowmem instance could conceivably take longer than 5 minutes to become available. I think, for "this should run until I cancel it", we should use `sleep 3600`.